### PR TITLE
solver function argument checking

### PR DESCRIPTION
**Pull request by daineAMD**
Updating info codes in solver-based functions (based off of solver argument checking code). Adding bad_arg tests for solver functions to ensure we're testing the hipBLAS portion of the code.

getrs
-------
- A and ipiv can be nullptr if n == 0
- B can be nullptr if n == 0 || nrhs == 0
- info = -1 if transpose != N || T || C

gels
-----
- A can be nullptr if n == 0 || m == 0
- B can be nullptr if nrhs == 0 || (m == 0 && n == 0)
- info = -1 if transpose != N || T for real cases, or transpsoe != N || C for complex cases
- deviceInfo can be nullptr if batchCount == 0

Aside from deviceInfo in gels, batchCount == 0 doesn't skip any other checks, it looks like this is the case in argument checking in rocSOLVER.

### CODE REVIEW

File: clients/include/testing_gels_strided_batched.hpp
Comment (TorreZuk) [inline clients/include/testing_gels_strided_batched.hpp:59]: These should be hipblasStride, size_t, or ssize_t even though not required by your constants you want to use correct sized variables.

File: clients/include/testing_geqrf_strided_batched.hpp
Comment (TorreZuk) [inline clients/include/testing_geqrf_strided_batched.hpp:51]: Assume solver never tests negative stride either so fine to lose sign.

**Review by TorreZuk — COMMENTED** [review]
I'll leave it to solver team to approve, if it were up to me I would request changes that those info values have defines or enums.  The structure I approve.

File: clients/include/testing_gels.hpp
Comment (daineAMD) [reply clients/include/testing_gels.hpp:102]: The size values are arbitrary to check correct return value when ldb < M or ldb < N. Thought it would be ok to be explicit in these cases.
The EXPECT_EQ values all correspond to the index of the argument which has the invalid value. In this case the 8th argument (not counting handle) is ldb, which is invalid as it is less than M.

File: library/src/amd_detail/hipblas.cpp
Comment (daineAMD) [reply library/src/amd_detail/hipblas.cpp:15540]: See previous comment. Since the argument index is different for each function not sure how we could get rid of these cleanly, could add comments instead?

File: clients/include/testing_gels.hpp
Comment (TorreZuk) [reply clients/include/testing_gels.hpp:102]: LOL okay if it is that simple but is this documented in their API, I only read the doc on Gels function and didn't see any mention of negative info values.

Comment (TorreZuk) [reply clients/include/testing_gels.hpp:102]: Can you update the hipBLAS documentation to say this, and will see if solver team chimes in.  I see cublas has a function level comment: If info<0, the parameter in position -info is invalid.   That should be repeated for each function where it is true IMO

Comment (daineAMD) [reply clients/include/testing_gels.hpp:102]: hipBLAS mentions it in [hipblas.h documentation](https://github.com/ROCmSoftwarePlatform/hipBLAS/blob/develop/library/include/hipblas.h#L16818), although I suppose the wording is technically incorrect, I'll fix it in another commit. [cublas documentation](https://docs.nvidia.com/cuda/cublas/index.html#cublas-lt-t-gt-gelsBatched) as a reference.

Comment (TorreZuk) [reply clients/include/testing_gels.hpp:102]: Thanks, I was reading rocsolver doc.   Okay maybe the fix is only needed in rocsolver doc, the j-th argument maybe clear enough is it 0 based indexing but handle is not referred to.

**Review by TorreZuk — APPROVED** [review]
Okay with me

**Comment by daineAMD** [comment]
Did some testing with cuBLAS backend, there seems to be quite a bit of disparity here, especially with gelsBatched. I've added the differences between rocSOLVER and cuBLAS, with the assumption that the bad_arg code is correct (hopefully someone from rocSOLVER can verify) at the bottom of this comment.

As for my latest commit, I:
- disabled `A == nullptr` and `Ipiv == nullptr` tests for geqrfBatched with cuBLAS backend as it gives mem. access errors
- disabled `A == nullptr`, `ipiv == nullptr`, and `info == nullptr` tests for getrsBatched with cuBLAS backend for the same reasons
- disabled batchCount < 0 test for cuBLAS backend
- disabled all gelsBatched bad_arg tests for cuBLAS backend. There are a bunch of differences between rocSOLVER and cuBLAS backend here (assuming my tests are good tests). It seems to me that rocSOLVER/hipBLAS gives correct error codes/return status and cuBLAS gives incorrect ones.

It isn't really our problem to be testing out cuBLAS code, especially bad_arg testing, but I think it's worthwhile to keep so we can understand cuBLAS behaviour better and ensure our rocSOLVER/hipBLAS error statuses and whatnot are reasonable.

## Differences between cuBLAS/rocSOLVER backend for gelsBatched, getrsBatched, and geqrfBatched edge cases

### getrs
- nullptrs don’t give error codes the same as rocSOLVER, we get get an illegal memory access error.
- batchCount < 0 returns HIPBLAS_STATUS_EXECUTION_FAILED rather than HIPBLAS_STATUS_INVALID_VALUE, and we get info == 0

### gels
- nullptrs don’t give error codes the same as rocSOLVER, we get get an illegal memory access error.
- when op != HIPBLAS_OP_N, we get HIPBLAS_STATUS_NOT_SUPPORTED rather than HIPBLAS_STATUS_INVALID_VALUE, we also get info == 0 instead of info == -1. rocSOLVER supports transpose cases, we return invalid_value when we pass in conj-transpose to sgels/dgels or we pass in regular-transpose to cgels/zgels
- ldb < N is fine for cuBLAS whereas this gives INVALID_VALUE in rocSOLVER backend
- nrhs == 0 && B == nullptr gives memory access fault

- M < 0 gives info == -1 on cuBLAS, -2 on rocSOLVER
- nrhs < 0 gives info = -2 on cuBLAS, -4 on rocSOLVER
- lda < M gives info = -5 on cuBLAS, -6 on rocSOLVER
- ldb < M gives info = -7 on cuBLAS, -8 on rocSOLVER
- deviceInfo == nullptr gives info = 0 on cuBLAS, -10 on rocSOLVER. CuBLAS returns success, solver returns invalid_value
- batchCount < 0 gives info = -8 on cuBLAS, -11 on rocSOLVER

### geqrf
- nullptrs don’t give error codes the same as rocSOLVER, we get get an illegal memory access error. info == nullptr is caught on both platforms.

**Comment by daineAMD** [comment]
@jzuniga-amd @tfalders Wondering if I can get a review from someone from the solver team on this. Thanks! 😄

File: clients/include/testing_gels.hpp
Comment (tfalders) [reply clients/include/testing_gels.hpp:102]: I'm not really a fan of these explicit values, though it's more of a gut feeling than a logical objection. The approach to `lda` on line 85 is more to my liking. For `ldb`, perhaps we could instead do `max(M, N) - 1`.

File: clients/include/testing_getrs.hpp
Comment (tfalders) [inline clients/include/testing_getrs.hpp:163]: This has N = 0 instead of nrhs = 0. (Same for other getrs tests)

File: library/src/amd_detail/hipblas.cpp
Comment (tfalders) [inline library/src/amd_detail/hipblas.cpp:16757]: How about `max(m, n) * nrhs`?

**Review by tfalders — APPROVED** [review]
Looks good to me! I have just a couple minor comments, including not liking the "magic numbers" pointed out by Torre.

File: library/src/amd_detail/hipblas.cpp
Comment (daineAMD) [reply library/src/amd_detail/hipblas.cpp:16757]: I suppose that would work, I like this better though, it's more clear and doesn't depend on negatives getting handled earlier, it's also identical to some error checking code in rocSOLVER.
For now I just added comments in testing_gels.hpp (and (strided)_bathced) regarding the explicit values.

File: clients/include/testing_getrs.hpp
Comment (daineAMD) [reply clients/include/testing_getrs.hpp]: Fixed, will wait to see if it passes on cublas.

**Pull request merged**
